feature: Add support for Scala 3's Best Effort compilation

jchyb opened the pull request:
Related to feature request: https://github.com/scalameta/metals-feature-requests/issues/327
## Related PRs
* dotty PR: https://github.com/lampepfl/dotty/pull/17582
* bloop PR: https://github.com/scalacenter/bloop/pull/2049

The dotty one explains the reasoning behind the Best Effort compilation feature in detail, while the bloop one implements most of the logic of how projects are managed with the best effort options. All of them are needed for this feature to work here.

## Explanation
Best Effort is meant to be a set of Scala 3 compiler options that allow the errored, but typed programs, to be able to be serialized as a TASTy aligned format (Best Effort TASTy), to be able to reuse those trees in both the dependent projects, and in the presentation compiler. This will allow better handling of projects that do not compile, and all of the dependent ones.

Basically, we are able to obtain a separate directory from bloop, that holds Best Effort TASTy, and can use it in PC. At the same time, SemanticDB generation is left as-is, those are created in the build file generated by bloop. Most of the best effort handling is done by bloop, including correctly setting the classpath with best effort directories. Let me quickly explain a few situations concerning project compilation that could happen before, and what can happen now with these PRs (especially with how the bloop one works):
### Before:
* project fails compilation (we get nothing)
* project compiles („regular” artifacts for PC and goto def are generated) -> Project fails (previous artifacts can be reused)
### As of this PR:
* project fails compilation with an unsuccessful Best Effort compilation (we get nothing, like before)
* project compiles (regular artifacts are generated) -> project fails compilation with an unsuccessful Best Effort (previous artifacts can be reused, like before)
* project fails compilation with a successful Best Effort compilation (we get Best Effort artifacts) -> project compiles (we have only regular artifacts)
* project compiles (we have only regular artifacts) -> project fails (we have both the regular artifacts (for running), and best effort artifacts, having both on the classpath did not cause any significant problems in my tests, but that can obviously be changed)
* project fails compilation with a successful Best Effort compilation (we only have best effort artifacts) -> project fails compilation with an unsuccessful Best Effort compilation (we only have the previous best effort artifacts)

## Contents
Overall this PR is meant to add support for Best Effort compilation in metals, with a few additional considerations:
* Some unnecessary options are replaced from the bloop ones for the presentation compiler
* Previously, any project could be run from metals if semanticdb was able to be generated. This caused a somewhat weird UX for best effort. Now the build directory is explicitly checked for the existence of non META-INF files.
* Previously, in metals doctor, compilation would be set as correct if a compilation did not produce any errors. However, with best effort compilation, the compiler may not find any additional errors, beyond the ones from its dependencies (meaning that a project that cannot actually be compiled, could show up as one that was compiling). A note about that was added in the doctor.

Integration tests were also added (probably too few for now).

jchyb commented on metals/src/main/scala/scala/meta/internal/metals/CompilerConfiguration.scala:
```diff
@@ -162,7 +162,27 @@ class CompilerConfiguration(
+      val pcOptions =
+        options.filter(_ != "-Ybest-effort") :+ "-Ywith-best-effort-tasty"
+      val bestEffortDirs = scalaTarget.info
+        .getDependencies()
```
After rebasing, I'm having some issues with this scalaTarget.info.getDependencies() method. Supposing I have projects `lib` (no additional dependencies) and `app` (depends on `lib`), when I import the project and I open the `app` module first in vscode, metals creates a newCompiler here for it, but this getDependencies() method returns an empty list (if I first create the newCompiler for `lib` and then `dep`, it works). I wonder if it was the same earlier but I did not notice or if it was changed with the lazier classpath resolution...

tgodzik replied:
lazy classpath resolution should not change anything for non-bazel BSP server. At least that would be unexpected. Dependencies should be resolved for all targets at the same time, I don't really see why it would be empty.

jchyb replied:
Ah I found it, the problem turned out to be elsewhere (think I misread my logs). Sorry about that

jchyb commented:
I updated the bloop version (I did not realize before that bloop is published after every merge, that's so cool) and added some comments in a few places.
One thing that I should mention somewhere, is that while the best effort options in the compiler and bloop are able to work even with a file broken on a parser level, in metals I believe that scalameta is used for parsing and indexing, which is used in go-to-ref even before semanticdb is considered, so that does not work yet for those files (but presentation compiler does, so we do get completions etc.). I saw that there is a PR to use presentation compiler for go-to-ref in the future, so I did not try to fix anything here related to that scalameta parser usage

tgodzik commented:
We can try on the next version 1.5.18-47-e7577874-SNAPSHOT

jchyb commented:
I still did not manage to fix the MillServerCodeLensSuite test, but we might as well see how the rest are doing

tgodzik commented:
Curiously, MillServerCodeLensSuite doesn't use Bloop so everything should work the same.

jchyb commented:
Yes, those are broken by this PR, by the changes I made to checking if there are any artifacts in classpath before allowing to use code lens. I believe that everything caused by the bloop changes was fixed

tgodzik commented:
Looks like Mill issue might be unrelated. Somehow it started failing on main also :thinking:

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/Compilers.scala:
```diff
@@ -264,10 +269,13 @@ class Compilers(
-    if (isSuccessful) {
+    if (isSuccessful || isBestEffortCompilation) {
+      // // For best effort even if we get errors, we can generate new betasty files
+      // if (!isSuccessful && !isBestEffortCompilation) {
```
Should we remove the commented code?

tgodzik commented on tests/slow/src/test/scala/tests/best-effort/BestEffortCompilationSuite.scala:
```diff
@@ -0,0 +1,299 @@
+  val scalaVersion = "3.5.0-RC1"
+
+  // implemented in bloop
+  test("best-effort-error-diagnostics") {
```
I wonder if we could reuse some of the tests in CompilersLspSuite for outline? We would need to make it into a base suite and extends here

jchyb replied:
We can but error messages are different in Scala 2 and 3, I'll try to figure something out for that

jchyb replied:
Wait nvm there are no error diagnostics in outline, I misread the logs as always...

jchyb replied:
Done, I did have to have them test different diagnostics for scala 3 and I also had to add saving after changes, before testing completions (but for scala-2s outline they should work identically). I could not make one of the test work (not sure why), so I moved it out of the base suite to an outline only one.

tgodzik reviewed (changes requested):
Thanks for working on this! I think we are almost finished. My plan for now is:
1. Release bloop with all the changes (it includes some other improvements)
2. Release metals
3. Merge this change so that we can have a period of testing.

tgodzik commented:
I already updated Bloop to 1.6.0, so you can rebase and we can merge after release

tgodzik commented on metals/src/main/scala/scala/meta/internal/metals/codelenses/RunTestCodeLens.scala:
```diff
@@ -92,37 +97,72 @@ final class RunTestCodeLens(
+   * but semanticDB and betasty (and thus the META_INF directory) may still be produced.
+   * We want to avoid creating run/test/debug code lens in those situations.
+   */
+  private def isNonMetaFileInClasspath(
```
Could we instead just check if the current build target is compiling? `diagnostics.hasCompilationErrors(buildId)` should be faster than looking up filesystem

jchyb replied:
That's true, but there is a case where a project may not produce any errors/diagnostics, but may be compiled based on betasty produced by a failing dependency. In that case we will still not get any artifacts there (outside of betasty/semanticdb), and trying to run will produce classpath errors. 

Though now that I think about it, we could check diagnostics of every dependency of a project to guess if a project was compiled using betasty. I'll try doing that

jchyb replied:
After reviewing my changes just now, I realized that I do the same thing in doctor, not sure why I didn't think to do this here too until now 😅

tgodzik commented on tests/unit/src/test/scala/tests/OutlineLspSuite.scala:
```diff
@@ -0,0 +1,77 @@
+  override val saveAfterChanges: Boolean = false
+  override val scala3Diagnostics = false
+
+  test("imports-for-non-compiling") {
```
Could we open an issue to look into it later?

jchyb replied:
Done (hopefully ok): https://github.com/scalameta/metals/issues/6593

tgodzik commented on tests/unit/src/test/scala/tests/best-effort/BestEffortCompilationSuite.scala:
```diff
@@ -0,0 +1,303 @@
+    } yield ()
+  }
+
+  // Here we test wheater completion symbols persist after removing
```
```suggestion
  // Here we test whether completion symbols persist after removing
```

tgodzik commented:
Looks like the tests are failing now, but this is expected since the test checks if the code lenses are kept when there are errors and maybe they shouldn't? At least not when the code is still not saved. I think we did it to avoid the code lenses popping in and out of existence thus moving the lines up and down.

Maybe it's enough if it doesn't do that `onChange` ? Since after saving with errors we would running some unknown version. 

Let's change `keep-after-error` to do onChange for the error and if you could run scalafix via `sbt scalafixAll` that would be awesome.

tgodzik reviewed (approved):
LGTM!

jchyb commented:
Thank you so much for overseeing this!

kpodsiad commented:
@jchyb @tgodzik I see PR for Bloop but I don't see sbt being mentioned anywhere. Does it mean that best effort compilation is only supported through Bloop?

tgodzik commented:
Currently yes,  it is a bit more complicated to set up than just adding a flag unfortunately
